internal/devconfig/ast.go

```diff
@@ -0,0 +1,282 @@
+type configAST struct {
+	root hujson.Value
+}
+
+// parseConfig parses the bytes of a devbox.json and returns a syntax tree.
+func parseConfig(b []byte) (*configAST, error) {
+	root, err := hujson.Parse(b)
+	if err != nil {
+		return nil, err
+	}
+	return &configAST{root: root}, nil
+}
+
+// packagesField gets the "packages" field, initializing it if necessary. The
+// member value will either be an array of strings or an object. When it's an
+// object, the keys will always be package names and the values will be a
+// string or another object. Examples are:
+//
+//   - {"packages": ["go", "hello"]}
+//   - {"packages": {"go": "1.20", "hello: {"platforms": ["aarch64-darwin"]}}}
+//
+// When migrate is true, the packages value will be migrated from the legacy
+// array format to the object format. For example, the array:
+//
+//	["go@latest", "hello"]
+//
+// will become:
+//
+//	{
+//		"go": "latest",
+//		"hello": ""
+//	}
+func (c *configAST) packagesField(migrate bool) *hujson.ObjectMember {
+	rootObject := c.root.Value.(*hujson.Object)
+	i := c.memberIndex(rootObject, "packages")
+	if i != -1 {
+		switch rootObject.Members[i].Value.Value.Kind() {
+		case '[':
+			if migrate {
+				c.migratePackagesArray(&rootObject.Members[i].Value)
+				c.root.Format()
+			}
+		case 'n':
+			// Initialize a null packages field to an empty object.
+			rootObject.Members[i].Value.Value = &hujson.Object{
+				AfterExtra: []byte{'\n'},
+			}
+			c.root.Format()
+		}
+		return &rootObject.Members[i]
+	}
+
+	// Add a packages field to the root config object and initialize it with
+	// an empty object.
+	rootObject.Members = append(rootObject.Members, hujson.ObjectMember{
+		Name: hujson.Value{
+			Value:       hujson.String("packages"),
+			BeforeExtra: []byte{'\n'},
+		},
+		Value: hujson.Value{Value: &hujson.Object{}},
+	})
+	c.root.Format()
+	return &rootObject.Members[len(rootObject.Members)-1]
+}
+
+// appendPackage appends a package to the packages field.
+func (c *configAST) appendPackage(name, version string) {
+	pkgs := c.packagesField(false)
+	switch val := pkgs.Value.Value.(type) {
+	case *hujson.Object:
+		c.appendPackageMember(val, name, version)
+	case *hujson.Array:
+		c.appendPackageElement(val, joinNameVersion(name, version))
+	default:
+		panic("packages field must be an object or array")
+	}
+
+	// Ensure the packages field is on its own line.
+	if !slices.Contains(pkgs.Name.BeforeExtra, '\n') {
+		pkgs.Name.BeforeExtra = append(pkgs.Name.BeforeExtra, '\n')
+	}
+	c.root.Format()
+}
+
+func (c *configAST) appendPackageMember(pkgs *hujson.Object, name, version string) {
+	i := c.memberIndex(pkgs, name)
+	if i != -1 {
+		return
+	}
+
+	// Add a new member to the packages object with the package name and
+	// version.
+	pkgs.Members = append(pkgs.Members, hujson.ObjectMember{
+		Name:  hujson.Value{Value: hujson.String(name), BeforeExtra: []byte{'\n'}},
```

Comment:
what does the "Extra" represent in BeforeExtra? Trying to understand why it is used.

Reply:
The extra fields track the whitespace, commas, and comments before/after/inside JSON values. So in ` /* comment */    "foo"` the foo string would have ` /* comment */    ` in its BeforeExtra field.